### PR TITLE
Update KhouryDataParams course types

### LINKED ISSUE #743: [Login Flow] -- Update User Course Registration
- [x] Update the Khoury Data Export types used in the login controller. Updated types can be found in Juan's doc. This is for Professors, TAs and Students. 
- [x] The Course Section Mapping Model needs to be updated. Replace `Generic Course Name` with `CRN`. 
- [x] Update the Login Course Service: update particularly how we create user courses using these new types.

### PR DESCRIPTION
**Pull request by nicoledanuwidjaja**
# Description
As per new login flow, updated course types in `KhouryDataParams` for new `KhouryCourse` and `KhouryProfCourse` objects. New course types now support courses identified by `crn`.

This PR will break existing data flow!

Closes #743 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Updated tests in `login.integration.ts`.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code where needed 
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by nicoledanuwidjaja** [comment]
Cypress tests are failing:
```
Failed to compile.

./components/Settings/CourseOverrideSettings.tsx:77:30
Type error: Property 'toggleSelfEnroll' does not exist on type '{ get: (courseId: number) => Promise<GetCourseResponse>; updateCalendar: (courseId: number) => Promise<unknown>; getCourseOverrides: (courseId: number) => Promise<...>; addOverride: (courseId: number, params: UpdateCourseOverrideBody) => Promise<...>; deleteOverride: (courseId: number, params: UpdateCourseOverrideBo...'.

  75 |         <Switch
  76 |           onChange={async () => {
> 77 |             await API.course.toggleSelfEnroll(courseId);
     |                              ^
  78 |             mutateCourse();
  79 |           }}
  80 |           defaultChecked={course?.selfEnroll}

File: packages/server/src/login/course-section-mapping.entity.ts [hunk 1]
```diff
@@ -13,12 +13,8 @@ export class CourseSectionMappingModel extends BaseEntity {
   @PrimaryGeneratedColumn()
   id: number;
 
-  // This is the course name that is sent to us from the khoury amin backend
   @Column()
-  genericCourseName: string;
-
-  @Column()
-  section: number;
+  crn: number;
```
Comment (dfarooq610) [inline packages/server/src/login/course-section-mapping.entity.ts:17]: Y'all need to make a migration for this

File: packages/server/src/login/login-course.service.ts [hunk 2]
```diff
@@ -86,13 +86,10 @@ export class LoginCourseService {
...
 
-  public async courseSectionToCourse(
-    courseName: string,
-    courseSection: number,
-  ): Promise<CourseModel> {
+  public async courseCRNToCourse(courseCRN: number): Promise<CourseModel> {
     const courseSectionModel = (
       await CourseSectionMappingModel.find({
-        where: { genericCourseName: courseName, section: courseSection },
+        where: { crn: courseCRN },
```
Comment (dfarooq610) [inline packages/server/src/login/login-course.service.ts:92]: This works because old courses _should_ be disabled, but that is through personal maintenance (running scripts and whatnot)

@IrisLiu-00 should we validate the semester name just to double check?

**Review by dfarooq610 — CHANGES_REQUESTED** [review]

File: packages/server/src/login/login-course.service.ts [hunk 3]
```diff
@@ -31,38 +31,38 @@ export class LoginCourseService {
...
+      // loop through section groups
+      for (const c of info.prof_courses) {
+        // loop through CRNs to create courses
+        for (const courseCRN of c.crns) {
+          const courseMappings = (
+            await CourseSectionMappingModel.find({
+              where: { crn: courseCRN }, // TODO: Add semester support
+              relations: ['course'],
+            })
+          ).filter((cm) => cm.course.enabled);
```
Comment (IrisLiu-00) [inline packages/server/src/login/login-course.service.ts:56]: It looks like the logic here is the same as what's happening in `courseCrnToCourse`, so you can probably just call that function. Also I believe you only need to inspect one of the CRNs here - all of them should map to the same CourseModel since they're all for the same section group.

File: packages/server/src/login/login-course.service.ts (hunk 2 above)
Comment (IrisLiu-00) [reply packages/server/src/login/login-course.service.ts:92]: Yeah I think this function should also take in a semester, since CRNs can potentially repeat between semesters. The MappingModel doesn't have a semester attribute but the CourseModel does - you'll have to do some joins and filters here. You'll also need to map the semester string from the admin data to our SemesterModels :P

File: packages/api-client/index.ts [hunk 4]
```diff
@@ -15,12 +15,12 @@ import {
   GetQueueResponse,
   GetReleaseNotesResponse,
   GroupQuestionsParams,
-  GetSelfEnrollResponse,
+  //GetSelfEnrollResponse,
```
Comment (IrisLiu-00) [inline packages/api-client/index.ts:18]: I'm guessing this being commented out is why Vercel/Cypress/linting/tests are failing on self enrollment stuff? Why was this commented out? Same for the TACheckinTimesResponse

**Review by IrisLiu-00 — CHANGES_REQUESTED** [review]
Looking good so far, put in some questions and requested changes

It would be nice to test out the login endpoint using postman or something, if you guys havent already - I'll also do a bit of manual testing before I give final PR approval.

File: packages/server/src/login/course-section-mapping.entity.ts (hunk 1 above)
Comment (brianyyang) [reply packages/server/src/login/course-section-mapping.entity.ts:17]: The migration file has been committed 🎉

File: packages/server/src/login/login-course.service.ts (hunk 2 above)
Comment (brianyyang) [reply packages/server/src/login/login-course.service.ts:92]: That makes sense, will attempt this now

File: packages/api-client/index.ts (hunk 4 above)
Comment (brianyyang) [reply packages/api-client/index.ts:18]: I think it was a linting issue once we commented out the previous registration form

File: packages/server/src/login/login-course.service.ts (hunk 2 above)
Comment (dfarooq610) [reply packages/server/src/login/login-course.service.ts:92]: For some more information: the Admin data will probably be formatted something like `Summer_Full 21`. 

Internally, we use the full year 2021 to prevent Y21K, so you make need to do a small bit of logic on the year there to make sure you query the right semester.

Comment (IrisLiu-00) [reply packages/server/src/login/login-course.service.ts:92]: ^ admin semester string format is in that notion doc outlining the admin post request

Comment (brianyyang) [reply packages/server/src/login/login-course.service.ts:92]: tyty

**Review by IrisLiu-00 — APPROVED** [review]
LGTM! I think you should be good to merge once you get another approval.

File: packages/server/src/login/login-course.service.ts [hunk 5]
```diff
@@ -31,38 +31,36 @@ export class LoginCourseService {
...
+            this.convertKhouryRole(c.role),
           );
-          userCourses.push(taCourse);
+          userCourses.push(userCourse);
+        }
+      } else {
+        // c is a KhouryProfCourse
+        // only need to inspect one of the CRNs from the list
+        // b/c they should all map to the same section group
+        if (c.crns.length !== 0) {
```
Comment (dfarooq610) [inline packages/server/src/login/login-course.service.ts:49]: Can we throw a 400 otherwise? So admin can get an error message if they give us bad data

File: packages/server/src/login/login-course.service.ts [hunk 6]
```diff
@@ -134,4 +136,37 @@ export class LoginCourseService {
...
+    let season = khourySeason;
+    switch (
+      khourySeason // summer sems are the only ones that are diff
+    ) {
+      case 'Summer I':
+        season = 'Summer_1';
+        break;
+      case 'Summer II':
+        season = 'Summer_2';
+        break;
```
Comment (dfarooq610) [inline packages/server/src/login/login-course.service.ts:168]: @jtavera235 just confirming one final time, the summer courses are formatted using Roman numerals like this, right?

Comment (dfarooq610) [reply packages/server/src/login/login-course.service.ts:168]: Yea just wanted to get your eyes on this parsing so we are 100% on the same page

File: packages/server/src/login/login-course.service.ts (hunk 5 above)
Comment (IrisLiu-00) [reply packages/server/src/login/login-course.service.ts:49]: In the interest of getting things moving, I think we can also come back to this when everything else is implemented.

File: packages/server/src/login/login-course.service.ts [hunk 7]
```diff
@@ -31,38 +31,36 @@ export class LoginCourseService {
...
+            course.id,
+            this.convertKhouryRole(c.role),
           );
-          userCourses.push(taCourse);
+          userCourses.push(userCourse);
+        }
+      } else {
+        // c is a KhouryProfCourse
+        // only need to inspect one of the CRNs from the list
+        // b/c they should all map to the same section group
```
Comment (jtavera235) [inline packages/server/src/login/login-course.service.ts:48]: This assumption is not necessarily true. An instructor can be part of multiple section groups as they can be teaching multiple different courses where each course belongs to a different section group so we would have to iterate through all the courses

File: packages/server/src/login/login-course.service.ts (hunk 6 above)
Comment (jtavera235) [reply packages/server/src/login/login-course.service.ts:168]: So this is on me but turns out the format of the semester is going to be different than originally planned. The admin page is going to return the semester code that appears on Banner. This is a 6 digit string where the first 4 digits are the year and the last 2 digits represent the semester with the semester mappings being:
10 - Fall
30 - Spring
40 _ sum1
50 - SumF
60 - Sum2
However when it is the Fall semester, it maps to the academic year. As an example:
202110 is Fall of 2020 - The academic year is 2021 but the current year of the fall semster is 2020.
202130 is Spring of 2021, etc. The weird rule only applies to Fall Semester

**Review by jtavera235 — CHANGES_REQUESTED** [review]
Really Great job! I left a comment with new changes in the semester format that I just learned about and a comment regarding an assumption that all courses will belong to the same section group. Other than that everything looks good

File: packages/server/src/login/login-course.service.ts (hunk 7 above)
Comment (dfarooq610) [reply packages/server/src/login/login-course.service.ts:48]: This is analyzing a single section group when iterating through all the prof's courses. So when trying to check if one of the professor's section groups is registered for the app, I think it should be fine to check one CRN inside of that. Either all or zero CRNs will be registered inside a Khoury professor course

Comment (jtavera235) [reply packages/server/src/login/login-course.service.ts:48]: Ohhh my bad yep all good!

**Review by jtavera235 — APPROVED** [review]

File: packages/server/src/login/login-course.service.ts [hunk 8]
```diff
@@ -142,31 +142,25 @@ export class LoginCourseService {
...
-      case 'Summer I':
-        season = 'Summer_1';
-        break;
-      case 'Summer II':
-        season = 'Summer_2';
-        break;
+    let year = Number(khourySemester.slice(0, 4));
+    const season = courseSeasonMap[khourySemester.slice(-2)];
+    // edge case for Fall semester
+    if (season == 'Fall') {
```
Comment (IrisLiu-00) [inline packages/server/src/login/login-course.service.ts:161]: while you're fixing tests - do triple equals, please!

**Review by dfarooq610 — APPROVED** [review]
